### PR TITLE
allow FIPS enabled systems to register successfully with puppet

### LINKED ISSUE #235: [RFE] Bootstrap should properly configure Puppet on a FIPS enabled system
See [bz1495215](https://bugzilla.redhat.com/show_bug.cgi?id=1495215). On FIPS enabled systems (where `/proc/sys/crypto/fips_enabled` == `1`) Bootstrap fails to run puppet (Due to puppet's default [signing algorithm](https://puppet.com/docs/puppet/5.3/configuration.html#digestalgorithm) being MD5, which isn't allowed in FIPS mode. 

Setting `digest_algorithm = sha256` in puppet.conf will allow a successful puppet run, but the signing algorithm must match on the Puppet Master.  (so this only works if both are set)

Questions/Thoughts:

- Do we default to setting the signing_algorithm to sha256 automatically when a FIPS enabled system is detected? 
- Do we make the user explicitly set 'FIPS mode' (which does the above)?
- Do we exit the script with an error?

### PR DESCRIPTION
**Pull request by sideangleside**
Fixes #235

### CODE REVIEW

File: bootstrap.py [hunk 1]
```diff
@@ -247,6 +247,18 @@ def get_puppet_version():
...
+def is_fips():
+    """
+    Checks to see if the system is FIPS enabled.
+    """
+    fipsfile = open("/proc/sys/crypto/fips_enabled", "r")
+    fipsStatus = fipsfile.read(1)
+    if fipsStatus == "1":
+        return True
+    else:
+        return False
```
Comment (akofink) [inline bootstrap.py]: Any reason not to do this?

```python
return fipsStatus == "1"
```

File: README.md [hunk 2]
```diff
@@ -480,6 +480,16 @@ Options:
                         separated
 ~~~
 
+# Additional Notes
+
+## FIPS support
+
+On systems with FIPS enabled (where `/proc/sys/crypto/fips_enabled` == `1`), algorithms such as MD5 are disallowed.
```
Comment (akofink) [inline README.md]: I think this should be '/proc/sys/crypto/fips_enabled == 1' (but using backticks, of course)

File: bootstrap.py (hunk 1 above)
Comment (sideangleside) [reply bootstrap.py]: It seemed more pythonic to return True/False in lieu of a value. But I can be convinced otherwise.

Comment (evgeni) [reply bootstrap.py]: `return fipsStatus == "1"` will return (True/False) just fine and is imho the pythonicst way to do it.

File: README.md [hunk 3]
```diff
@@ -480,6 +480,16 @@ Options:
...
+Bootstrap will configure `digest_algorithm = sha256` in puppet.conf to allow successful puppet runs.
+However, the signing algorithm **must** match on the Puppet Master. It is expected that the Puppet Masters are configured with
+the **same** algorithm.
```
Comment (evgeni) [inline README.md]: stray linebreak

File: bootstrap.py [hunk 4]
```diff
@@ -247,6 +247,18 @@ def get_puppet_version():
             return puppet_version
 
 
+def is_fips():
+    """
+    Checks to see if the system is FIPS enabled.
+    """
+    fipsfile = open("/proc/sys/crypto/fips_enabled", "r")
```
Comment (evgeni) [inline bootstrap.py]: imho this should be `fips_file`

File: bootstrap.py [hunk 5]
```diff
@@ -247,6 +247,18 @@ def get_puppet_version():
...
+    fipsStatus = fipsfile.read(1)
```
Comment (evgeni) [inline bootstrap.py]: and this `fips_status`, we're in snake_land here ;)

**Comment by sideangleside** [comment]
I've incorporated all of the above feedback into https://github.com/Katello/katello-client-bootstrap/pull/236/commits/97e1930388ff7bb130849fbdd724fd4e6462f8fa

File: bootstrap.py [hunk 6]
```diff
@@ -247,6 +247,16 @@ def get_puppet_version():
             return puppet_version
 
 
+def is_fips():
+    """
+    Checks to see if the system is FIPS enabled.
+    """
+    fips_file = open("/proc/sys/crypto/fips_enabled", "r")
+    fips_status = fips_file.read(1)
+    if fips_status == "1":
```
Comment (evgeni) [inline bootstrap.py]: nah, just always return it.

File: bootstrap.py [hunk 7]
```diff
@@ -247,6 +247,16 @@ def get_puppet_version():
             return puppet_version
 
 
+def is_fips():
+    """
+    Checks to see if the system is FIPS enabled.
+    """
+    fips_file = open("/proc/sys/crypto/fips_enabled", "r")
```
Comment (evgeni) [inline bootstrap.py:254]: I wonder if we should handle systems where this file does not exist. (like a CentOS container running on Debian).

File: bootstrap.py (hunk 6 above)
Comment (sideangleside) [reply bootstrap.py]: ah. got it now.

File: bootstrap.py (hunk 7 above)
Comment (sideangleside) [reply bootstrap.py:254]: Why wouldn't it exist?  (though if we are going to pretend bootstrap works on non-RHEL, we'll need a whole nother 5 pull requests)

File: bootstrap.py (hunk 1 above)
Comment (akofink) [reply bootstrap.py]: And, for those interested, it's the Ruby way also (though I'm struggling to find a link to back myself up).

File: bootstrap.py (hunk 7 above)
Comment (akofink) [reply bootstrap.py:254]: I think an EL/Fedora container running on non-EL/Fedora is the only way this could happen.

**Review by akofink — COMMENTED** [review]
LGTM after the one `fips_status == "1"` change is updated

File: bootstrap.py (hunk 7 above)
Comment (sideangleside) [reply bootstrap.py:254]: but the FIPS attribute is set at the kernel level. I suspect `/proc/sys/crypto/fips_enabled` exists on a Debian system. (Though I don't have one in front of me to confirm)

Comment (akofink) [reply bootstrap.py:254]: Yeah, it's dependent on the docker host. Every single OS image (centos, rhel, debian, alpine) has that file when I run docker on Fedora, but none of the images has it when I run on non-EL/Fedora, such as on OS X, for instance. To clarify, running a RHEL container does not even have that file if the container is running on an OS that doesn't have that file.

Comment (evgeni) [reply bootstrap.py:254]: Yeah, exactly. It's a kernel thing, so a container running on a non-EL/Fedora kernel will not have this.

It's not like I like registering containers anyways...

**Review by evgeni — APPROVED** [review]

**Review by akofink — APPROVED** [review]

**Pull request merged**